vertexai/src/ModelContent.cs

```diff
@@ -16,91 +16,260 @@
+      Name = name;
+      Args = new Dictionary<string, object>(args);
+    }
+
+    Dictionary<string, object> Part.ToJson() {
       throw new NotImplementedException();
     }
   }
 
+  /// <summary>
+  /// Result output from a function call.
+  ///
+  /// Contains a string representing the `FunctionDeclaration.name` and a structured JSON object
+  /// containing any output from the function is used as context to the model. This should contain the
+  /// result of a ``FunctionCallPart`` made based on model prediction.
+  /// </summary>
   public readonly struct FunctionResponsePart : Part {
     public string Name { get; }
     public IReadOnlyDictionary<string, object> Response { get; }
 
+    /// <summary>
+    /// Constructs a new `FunctionResponsePart`.
+    /// </summary>
+    /// <param name="name">The name of the function that was called.</param>
+    /// <param name="response">The function's response.</param>
     public FunctionResponsePart(string name, IDictionary<string, object> response) {
+      Name = name;
+      Response = new Dictionary<string, object>(response);
+    }
+
+    Dictionary<string, object> Part.ToJson() {
       throw new NotImplementedException();
     }
   }
+
+#endregion
+
+  internal Dictionary<string, object> ToJson() {
+    return new Dictionary<string, object>() {
+      ["role"] = Role,
+      ["parts"] = Parts.Select(p => p.ToJson()).ToList()
+    };
+  }
+
+  internal static ModelContent FromJson(Dictionary<string, object> jsonDict) {
+    // Parse the Role, which is required
+    string role;
+    if (jsonDict.TryGetValue("role", out object roleObj) && (roleObj is string roleStr)) {
+      role = roleStr;
+    } else {
+      throw new VertexAISerializationException("'content' missing valid 'role' field.");
+    }
+
+    // Parse the Parts, which are required
+    List<Part> partList = new();
+    if (jsonDict.TryGetValue("parts", out object partsObj) && (partsObj is List<object> partsObjList)) {
+      partList = partsObjList
+          .Select(o => o as Dictionary<string, object>)
+          .Where(d => d != null)
+          .Select(PartFromJson)
+          .ToList();
+    } else {
+      throw new VertexAISerializationException("'content' missing valid 'parts' field.");
+    }
+
+    return new ModelContent(role, partList);
+  }
+
+  private static Part PartFromJson(Dictionary<string, object> jsonDict) {
+    if (jsonDict.ContainsKey("text") && jsonDict["text"] is string text) {
+      return new TextPart(text);
+    } else if (jsonDict.ContainsKey("functionCall") && jsonDict["functionCall"] is Dictionary<string, object> functionCallDict) {
+      if (!functionCallDict.TryGetValue("name", out var name) || name is not string) {
+        throw new VertexAISerializationException("Invalid JSON format: 'name' is not a string.");
+      }
+      if (!functionCallDict.TryGetValue("name", out var args) || args is not Dictionary<string, object>) {
```

Comment:
Is this right? "name" should be in "args"?

Reply:
Whoops, should've been args, nice catch.